src/System.IO/tests/Stream/Stream.CopyToTests.cs

```diff
@@ -0,0 +1,338 @@
+        public void IfLengthIsLessThanOrEqualToPositionCopyToShouldStillBeCalledWithAPositiveBufferSize(long length, long position)
+        {
+            // Streams with their Lengths <= their Positions, e.g.
+            // new MemoryStream { Position = 3 }.SetLength(1)
+            // should still be called CopyTo{Async} on with a
+            // bufferSize of at least 1.
+
+            var baseStream = new DelegateStream(
+                canReadFunc: () => true,
+                canSeekFunc: () => true,
+                lengthFunc: () => length,
+                positionGetFunc: () => position,
+                readFunc: (buffer, offset, count) => 0);
+            var trackingStream = new CallTrackingStream(baseStream);
+
+            var dest = Stream.Null;
+            trackingStream.CopyTo(dest);
+
+            // No argument checking needed; see notes below
+        }
+
+        [Theory]
+        [MemberData(nameof(LengthIsLessThanOrEqualToPosition))]
+        public async void AsyncIfLengthIsLessThanOrEqualToPositionCopyToShouldStillBeCalledWithAPositiveBufferSize(long length, long position)
+        {
+            var baseStream = new DelegateStream(
+                canReadFunc: () => true,
+                canSeekFunc: () => true,
+                lengthFunc: () => length,
+                positionGetFunc: () => position,
+                readFunc: (buffer, offset, count) => 0);
+            var trackingStream = new CallTrackingStream(baseStream);
+
+            var dest = Stream.Null;
+            await trackingStream.CopyToAsync(dest);
+
+            Assert.Same(dest, trackingStream.CopyToAsyncDestination);
+            Assert.True(trackingStream.CopyToAsyncBufferSize > 0);
+            Assert.Equal(CancellationToken.None, trackingStream.CopyToAsyncCancellationToken);
+        }
+
+        [Theory]
+        [MemberData(nameof(LengthMinusPositionPositiveOverflows))]
+        public void IfLengthMinusPositionPositiveOverflowsBufferSizeShouldStillBePositive(long length, long position)
+        {
+            // The new implementation of Stream.CopyTo calculates the bytes left
+            // in the Stream by calling Length - Position. This can overflow to a
+            // negative number, so this tests that if that happens we don't send
+            // in a negative bufferSize.
+
+            var baseStream = new DelegateStream(
+                canReadFunc: () => true,
+                canSeekFunc: () => true,
+                lengthFunc: () => length,
+                positionGetFunc: () => position,
+                readFunc: (buffer, offset, count) => 0);
+            var trackingStream = new CallTrackingStream(baseStream);
+
+            var dest = Stream.Null;
+            trackingStream.CopyTo(dest);
+
+            // We don't need to do any argument checking here;
+            // if bufferSize was null CopyTo will throw an exception
+            // during argument validation
```

Comment:
@stephentoub Even if the argument validation is omitted, an exception is going to be thrown when `CopyTo` tries to allocate a negative-sized array. (I made a typo of null vs negative in the comment)